File: src/grammar.fs
module Grammar

//-----------------------------------------------------------------------------
// GRAMMAR

/// represents a legal list of states
type Rule<'State> = 'State list

/// represents a grammar
/// 
/// roostate is the original state
/// expand takes a state and returns all the possible expansions
/// eval takes a rule and returns a score (the target is to maximize the score)
/// 
/// WARNING:
/// expand should return the rule from the closest to a terminal state to the further
/// a terminal should return an empty array (since it has no expansion)
type Grammar<'State> =
   {
      rootState:'State
      expand: 'State -> Rule<'State> array
      evalFunction: Rule<'State> -> float
   }
   
   /// evaluate a rule and encapsulate it so that invalid result becomes None
   member this.eval rule =
      let result = this.evalFunction rule 
      if result = -infinity || System.Double.IsNaN result then None else Some result

   /// returns true if a state is terminal
   /// WARNING: this fucntion is as expensive as a call to grammar expand
   member this.isTerminal state =
      state |> this.expand |> Array.isEmpty
   
   /// builds a random formula
   member this.randomFormula maxlength =
      let rec expand maxlength result formula =
         match formula with 
         | [] -> 
            List.rev result
         | state::formula ->
            match this.expand state with 
            | [||] -> // terminal state
               expand maxlength (state::result) formula
            | rules -> // non terminal state
               let newStates = if maxlength <= 0 then rules.[0] else Random.element rules
               let maxlength = maxlength + 1 - List.length newStates
               expand maxlength result (newStates @ formula)
      expand maxlength [] [this.rootState]

//-----------------------------------------------------------------------------
// FUNCTIONS

/// takes an evaluation function, some [|variables, targets|] and produces the sum of squared error
/// this can be used as an evaluation function when doing regression
let regression evaluation data formula : float =
   // norm2 error
   let error2 (output, inputs) =
      let result = evaluation inputs formula
      abs (result - output) // norm1
      //(result - output)*(result - output) // norm2
   // compute the sum of norm two errors
   - (Array.sumBy error2 data) / (float data.Length)

/// takes a constructor and an array of values and output an array or rules
/// useful to write a grammar quicker
let rules constructor values =
   Array.map (fun v -> [constructor v]) values

File: src/grammar.rs
/*
   an implementation of the, very simple, 2019 grammar
*/

/// represents a state
#[derive(Copy, Clone, Debug)]
pub enum State
{
   Expr,
   One,
   Add,
   Mul
}

/// represents the root of a formula
pub static ROOTSTATE: State = State::Expr;

/// expands a state into potential substitution rules
/// an empty vector represents a terminal state  :there is no rule associated with it
pub fn expand(state: State) -> Vec<Vec<State>>
{
   match state
   {
      State::Expr => vec![vec![State::One], vec![State::Add, State::Expr, State::Expr], vec![State::Mul,
                                                                                             State::Expr,
                                                                                             State::Expr]],
      _ => vec![]
   }
}

/// computes a formula
fn compute(formula: &[State]) -> i64
{
   /// computes the first element of the formula and returns its value followed with any leftover
   fn compute_rec(formula: &[State]) -> (i64, &[State])
   {
      match formula
      {
         [State::One, formula..] => (1, formula),
         [State::Add, formula..] =>
         {
            let (x, formula) = compute_rec(formula);
            let (y, formula) = compute_rec(formula);
            (x + y, formula)
         }
         [State::Mul, formula..] =>
         {
            let (x, formula) = compute_rec(formula);
            let (y, formula) = compute_rec(formula);
            (x * y, formula)
         }
         [uncomputable, ..] => panic!("Tried to compute a non terminal state : {:?}", uncomputable),
         [] => panic!("Tried to compute the empty formula.")
      }
   }
   // checks wether there is any leftover
   match compute_rec(formula)
   {
      (result, []) => result,
      (_, leftover) => panic!("There are some leftover states : {:?} => {:?}", formula, leftover)
   }
}

/// evaluates a formula
pub fn evaluate(formula: &[State]) -> Option<f64>
{
   let value = compute(formula);
   let score = (2019 - value) as f64;
   Some(score)
}